feat: refined ux update a taxonomy by downloading and uploading  (TEMP)

rpenido opened the pull request:
## Description
This PR improves the import tags functionality for existing taxonomies implemented at https://github.com/openedx/frontend-app-course-authoring/pull/675.

## Supporting Information
* Closes:
  * https://github.com/openedx/modular-learning/issues/126
* Depends on:
  * https://github.com/openedx/frontend-app-course-authoring/pull/675
  * https://github.com/openedx/openedx-learning/pull/130

## Testing instructions
### Re-import tags to an existing taxonomy
1. Have an import taxonomy template ready to be imported. You could get one from below:
    - [template.csv](https://github.com/openedx/frontend-app-course-authoring/files/13322469/template.csv)
    - [template.json](https://github.com/openedx/frontend-app-course-authoring/files/13322474/template.json)
1. Go to the taxonomy list page (http://localhost:2001/taxonomies)
1. Click on the `Re-import` menu from an existing taxonomy card
1. You will be prompted with the first page, where you can export the current taxonomy in CVS or JSON format. Check if both buttons export the taxonomy in the desired format:
![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/41a1ee0f-40d3-444b-a593-643d2ca17988)
1. Click on `Next`
1.  Will you have to select a CSV or JSON format file to import. You could do this clicking on the component or using drag and drop. Only .csv and .json files should be accepted.
![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/a85ce970-127e-45f9-8532-83cf89259d5c)
1. Click on `Import` to go to the next step.
1. This page will show the preview actions for this import. Check it and click on `Continue`. If the file shows no differences, the `Continue` button will be disabled. 
![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/2798bcbc-b2d0-446e-8fc8-ae917e3f5257)
1. A new confirmation windows will be show. Clicking on `Yes, import file` should import the tags, replacing the old ones.
![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/810964b4-4c7d-43a0-a2a1-4fe888ecc09f)
1. The Tag List should be automatically updated.

____
Private-ref:
 - [FAL-3539](https://tasks.opencraft.com/browse/FAL-3539)

rpenido committed:
feat: refined ux update a taxonomy by downloading and uploading

rpenido committed:
fix: eslint

rpenido committed:
fix: design adjustments

rpenido committed:
fix: more design adjustments

rpenido committed:
test: fix tests

rpenido committed:
test: fix api test

ChrisChV commented on src/taxonomy/import-tags/ImportTagsWizard.jsx:
```diff
@@ -0,0 +1,335 @@
+import PropTypes from 'prop-types';
+import { useState } from 'react';
+
+import { getFileSizeToClosestByte } from '../../files-and-videos/generic/utils'; // ToDo: Check best approach
```
@rpenido Maybe you need to extract this to `src/utils`

rpenido replied:
Done 6a55c6f3c6895a9b5b96b1f8ae9e30c3fdff06c

ChrisChV commented on src/taxonomy/import-tags/data/api.js:
```diff
@@ -40,19 +50,62 @@ export async function importNewTaxonomy(taxonomyName, taxonomyDescription, file)
+    },
+    onSuccess: (_data, variables) => {
+      queryClient.invalidateQueries({
+        queryKey: ['tagList', variables.taxonomyId],
```
@rpenido I think also you need to invalidate the `taxonomyDetail` query

rpenido replied:
Thank you @ChrisChV! I missed the tag count on Taxonomy.
Done 111904180de7951ec038b272cae5711c62ce8962

rpenido committed:
refactor: move getFileSizeToClosestByte to root utils

rpenido committed:
fix: invalidate taxonomyDetail query

rpenido committed:
test: add testing-library/react-hooks and fix testing

rpenido committed:
fix: rename files

rpenido committed:
test: fix api test

rpenido committed:
test: check invalidateQueries

rpenido committed:
test: improve LoadingButton tests

rpenido committed:
test: fix LoadingButton test

rpenido committed:
test: add ImportTagsWizard tests

rpenido committed:
test: more ImportTagsWizard tests

rpenido committed:
fix: types

rpenido committed:
feat: add typings from code review

Co-authored-by: Braden MacDonald <mail@bradenm.com>

rpenido committed:
refactor: typings and return of taxonomy import api

rpenido committed:
fix: type import

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
fix: TaxonomyContext types

rpenido committed:
feat: add toast and error message

rpenido committed:
fix: layout alert message

rpenido commented on src/taxonomy/import-tags/data/api.test.js:
```diff
@@ -1,48 +0,0 @@
-import MockAdapter from 'axios-mock-adapter';
```
This was renamed to `.jsx` to allow <QueryClientProvider> component wrapper

rpenido commented on src/taxonomy/import-tags/data/utils.js:
```diff
@@ -89,32 +90,3 @@ export const importTaxonomy = async (intl) => {
       console.error(error.response);
     });
 };
-
```
This function was replaced by the ImportTagsWizard

rpenido commented on src/files-and-videos/generic/utils.js:
```diff
@@ -1,23 +1,4 @@
-export const getFileSizeToClosestByte = (fileSize, numberOfDivides = 0) => {
```
This function was moved to /src/utils.js

rpenido committed:
fix: types and lint

rpenido committed:
fix: lint

ChrisChV commented:
@rpenido I can't continue with the import. Maybe I'm missing something?

https://www.loom.com/share/4bf7b266cda4407f914b7100a39fe639?sid=ff2bc3b2-78b1-4c8a-b088-1ba793a8a77b

ChrisChV commented on src/taxonomy/import-tags/ImportTagsWizard.jsx:
```diff
@@ -0,0 +1,374 @@
+        <div>
+          {!file ? (
+            <Dropzone
+              style={{ height: '200px' }}
```
@rpenido It seems like you are using this style in several places. maybe it's better to [create a utility](https://github.com/openedx/frontend-app-course-authoring/commit/dab9d8a01a0389826e989252b112c3b70869e251#diff-c2d8998567caf1b28fc98e349d9fdbe0b95a3e4ea15938b79b7ce147b90b9cbdR5). Ref: https://github.com/openedx/frontend-app-course-authoring/pull/723#discussion_r1425040465

rpenido replied:
Thank you! FIxed here: 2d3abe8542acc16d49bbceb59aed9e1fe82e5bd7

ChrisChV commented:
Also we need to improve the error messages here, but it can be on a follow-up task CC @bradenmacdonald 

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/11827305/885e7700-d2dd-4779-93ab-f81f1be3de6c)

rpenido committed:
fix: error message

rpenido commented:
> @rpenido I can't continue with the import. Maybe I'm missing something?
> 
> https://www.loom.com/share/4bf7b266cda4407f914b7100a39fe639?sid=ff2bc3b2-78b1-4c8a-b088-1ba793a8a77b

Thank you Chris! I fixed a bug in the error alert message here: ad69e11344172cc313f91c16060af76e8e84f806

But I think you also found a bug in our import function. The step `#30` is showing this:
```
[2023-12-18 16:51:19]  #30: Delete tag (external_id=None) [Started]
[2023-12-18 16:51:19] MultipleObjectsReturned('get() returned more than one Tag -- it returned more than 20!') 
```

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
fix: use scss utility

rpenido committed:
style: add PR in comment

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
fix: missing }

rpenido commented:
> I can fix it if you want, I have time for it

I would appreciate that @ChrisChV !

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
fix: import taxonomy type

rpenido committed:
revert: changes in package-lock.json

ChrisChV commented:
> > I can fix it if you want, I have time for it
> 
> I would appreciate that @ChrisChV !
> 
> If this is blocking for testing here, the current code will work reimporting any taxonomy created via Import.

@rpenido I fixed this bug on this PR: https://github.com/openedx/openedx-learning/pull/135
I need to add tests, but you can review it and test it with your PR

rpenido committed:
refactor: rename close -> onClose

rpenido committed:
test: change to getByText

rpenido commented:
> @rpenido I fixed this bug on this PR: [openedx/openedx-learning#135](https://github.com/openedx/openedx-learning/pull/135) I need to add tests, but you can review it and test it with your PR

I will check it tomorrow. Thank you @ChrisChV !

rpenido committed:
Merge branch 'rpenido/fal-3532-import-taxonomy' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

renovate[bot] committed:
fix(deps): update dependency @edx/frontend-lib-content-components to v1.177.9 (#762)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

ChrisChV reviewed (approved):
@rpenido :+1:
- [X] I tested this: I followed the testing instructions
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] I tested that the UI can be used with a keyboard only (tab order, keyboard controls).
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

rpenido committed:
fix: add a div over the dialog to prevent the user to interact while processing

rpenido committed:
fix: add title to dialog to remove console warning

rpenido committed:
fix: add title property to Stepper.Step to fix warning

rpenido committed:
fix: remove warning when toastMessage = null

yusuf-musleh committed:
feat: Add filter taxonomies by org (#755)

This implements filtering taxonomies on the taxonomy list page by selecting organization name, all taxonomies, or unassigned taxonomies.

KristinAoki committed:
fix: uploading progress percentage (#763)

ArturGaspar committed:
feat: error page on invalid course key (#761)

ayesha-waris committed:
chore: removed reported_content_email_notifications_flag dependency (#775)

rpenido committed:
feat: add import taxonomy feature [FC-0036] (#675)

This change adds a new button in the Taxonomy List to allow users to create new taxonomies by importing a CSV/JSON file.

yusuf-musleh committed:
feat: Add v2 lib components content tags support (#771)

Add support for fetching content data for Library V2 components in content tags drawer.

rpenido committed:
Merge branch 'master' into rpenido/fal-3539-refined-ux-update-a-taxonomy-by-downloading-and-uploading

rpenido committed:
fix: merging errors

rpenido commented:
Closed in favor of https://github.com/openedx/frontend-app-course-authoring/pull/732
